### ISSUE TITLE
Rewrite metadata classes bootstrapping in Level1 "pnext"

### ISSUE THREAD

**Issue by per-samuelsson** [Starcounter/Starcounter#3029]
Extends #2728.

**Comment by k-rus** [comment]
I noticed that I forgot to commit one change in metaschema.json. Now it is pushed to pnext.

**Comment by per-samuelsson** [comment]
> Now it is pushed to pnext of Level0.

Super.

**Comment by per-samuelsson** [comment]
@k-rus 

Now that I switch to **pnext** again, do I have to regenerate the VS-env, or is that just a one-time operation? I mean, will the one I generated last time always be up-to-date for example if files are added?

**Comment by k-rus** [comment]
> Now that I switch to pnext again, do I have to regenerate the VS-env, or is that just a one-time operation? I mean, will the one I generated last time always be up-to-date for example if files are added?

I can answer only from my experience: if I am not sure, then I always run cmake to create the solution and project files and then build Level0. However, it might working by just building Level0, since it seems that the projects are re-generated if needed. But I am not sure if it always work or not.

**Comment by per-samuelsson** [comment]
> But I am not sure if it always work or not.

OK, thanks. I'll just go with trial and error then. :smile:

**Comment by per-samuelsson** [comment]
I've got things to compile finally (by deleting everything and building totally from scratch), but starting the server fails on a error in the gateway, I think its a access violation, and I just cant seem to get around it, so right now the progress of this is 0.

**Comment by per-samuelsson** [comment]
```
staradmin list log

[09:15:42, Error, Starcounter (service)]
ScErrNetworkGatewayTerminated (SCERR10021): The server detected that the network gateway executable (scnetworkgateway.exe) unexpectedly terminated.

[09:15:42, Error, Starcounter (service)]
Process exitcode: -1073741819

[09:15:42, Critical, Starcounter (administrator)]
System.Exception: ScErrUnspecified (SCERR999): An unspecified error caused the operation to fail. System.Net.Sockets.SocketException (0x80004005): An existing connection was forcibly closed by the remote host
   at System.Net.Sockets.Socket.Receive(Byte[] buffer, Int32 offset, Int32 size, SocketFlags socketFlags)
   at Starcounter.NodeTask.PerformSyncRequest() in C:\Users\Per\Git\Starcounter\Develop\Level1\src\Starcounter.Internal\Rest\NodeTask.cs:line 545
Version: 2.0.0.
Help page: https://github.com/Starcounter/Starcounter/wiki/SCERR999.
   at Starcounter.GatewayHandlers.RegisterUriHandlerNative(UInt16 port, String appName, String originalUriInfo, String processedUriInfo, Byte[] nativeParamTypes, UInt16 managedHandlerIndex, UInt64& handlerInfo) in C:\Users\Per\Git\Starcounter\Develop\Level1\src\Starcounter\Apps\GatewayHandlers.cs:line 516
   at Starcounter.Rest.UriHandlersManager.RegisterUriHandler(UInt16 port, String originalUriInfo, String processedUriInfo, Byte[] nativeParamTypes, Type messageType, Func`4 wrappedDelegate, NetworkProtocolType protoType, HandlerOptions ho) in C:\Users\Per\Git\Starcounter\Develop\Level1\src\Starcounter.Rest\UriHandlersManager.cs:line 654
   at Starcounter.Rest.UriManagedHandlersCodegen.RegisterDelegate(UInt16 port, String methodSpaceUri, MethodInfo userDelegateInfo, Expression delegExpr, NetworkProtocolType protoType, HandlerOptions ho) in C:\Users\Per\Git\Starcounter\Develop\Level1\src\Starcounter.Rest\UriManagedHandlersCodegen.cs:line 609
   at Starcounter.Rest.UriManagedHandlersCodegen.GenerateParsingDelegate[T1,T2](UInt16 port, String methodSpaceUri, Func`3 userDelegate, HandlerOptions ho, NetworkProtocolType protoType) in C:\Users\Per\Git\Starcounter\Develop\Level1\src\Starcounter.Rest\UriManagedHandlersCodegen.cs:line 876
   at Starcounter.Rest.RestRegistrationProxy.RegisterHandler[T1,T2](UInt16 port, String methodSpaceUri, Func`3 handler, HandlerOptions ho) in C:\Users\Per\Git\Starcounter\Develop\Level1\src\Starcounter.Rest\RestInitialization.cs:line 78
   at Starcounter.Handle.PATCH[T1,T2](UInt16 port, String uriTemplate, Func`3 handler, HandlerOptions ho) in C:\Users\Per\Git\Starcounter\Develop\Level1\src\Starcounter.Internal\Handle.cs:line 835
   at Starcounter.Internal.PuppetRestHandler.RegisterJsonPatchHandlers(UInt16 port) in C:\Users\Per\Git\Starcounter\Develop\Level1\src\Starcounter.Apps.JsonPatch\BuiltInRestHandlers\PuppetRestHandler.cs:line 83
   at Starcounter.Internal.AppsBootstrapper.InitAppsBootstrapper(Byte numSchedulers, UInt16 defaultUserHttpPort, UInt16 defaultSystemHttpPort, UInt32 sessionTimeoutMinutes, String dbName, Boolean noNetworkGateway) in C:\Users\Per\Git\Starcounter\Develop\Level1\src\Starcounter.Apps.JsonPatch\AppsBootstrapper.cs:line 107
   at StarcounterInternal.Bootstrap.Control.Setup(String[] args) in C:\Users\Per\Git\Starcounter\Develop\Level1\src\Starcounter.Bootstrap\Control.cs:line 189
   at StarcounterInternal.Bootstrap.Control.Main(String[] args) in C:\Users\Per\Git\Starcounter\Develop\Level1\src\Starcounter.Bootstrap\Control.cs:line 54
HResult=-2146233088
HelpLink=https://github.com/Starcounter/Starcounter/wiki/SCERR999
```

**Comment by k-rus** [comment]
I don't know the reason, but it might be because branch merge2728 is not synchronised with latest changes in pnext. I often experienced such kind of problem.

**Comment by per-samuelsson** [comment]
>  [...] it might be because branch merge2728 is not synchronised with latest changes in pnext. I often experienced such kind of problem.

I've merged `pnext` into this branch you pointed me to, so its something else.

**Comment by per-samuelsson** [comment]
Looks like I might be experiencing something like #2962. You regularly merge the fixes in `develop` into `pnext`, @k-rus?

**Comment by k-rus** [comment]
> You regularly merge the fixes in develop into pnext, @k-rus?

No. Usually @eriohl was doing such merge, since he knows better about important changes in Level0.
So far I haven't done any merges of develop into pnext.
We haven't defined any procedures and responsibilities how to move forward with merges. So it is completely ad hoc on need basis. I guess now conflicting changes might fall to different developers as several people works with managed pnext and changes in develop happens in parallel.

You can try to merge and do it first in pnext and if works push it into remote pnext, before continuing on this issue.

**Comment by per-samuelsson** [comment]
I have now tried running things in isolation, and if I first start the IPC monitor and then setup a debugging session and start the gateway there (using the same arguments as those used by `scservice`), I end up with this:

![image](https://cloud.githubusercontent.com/assets/2902045/11030212/e08a990c-86cc-11e5-90c9-35d4d0e37f3d.png)

So its clearly not the same thing as #2962, but instead something (I guess) with my environment that gets fucked up when swiching `develop` <--> `pnext` contexts. Looks like something with CLANG.

Will have to further investigate this.

**Comment by k-rus** [comment]
I merged pnextMerge2728 branch and pnext locally and QueryProcessingTest works until bootstrap error. @per-samuelsson Should I push the merge into pnextMerge2728 or would you prefer to do it yourself when you solve the environment problem?

**Comment by per-samuelsson** [comment]
> Should I push the merge into pnextMerge2728 or would you prefer to do it yourself when you solve the environment problem?

You can push it, no problem. My problem with my env is not really related.

**Comment by k-rus** [comment]
> Should I push the merge into pnextMerge2728 or would you prefer to do it yourself when you solve the environment problem?
> You can push it, no problem.

Done

**Comment by per-samuelsson** [comment]
So, I've spent some time now trying to debug the gateway and how it use `scllvm`, which seem to be the originator of the access violation. I still havent found the exact cause, but its clear that its happening when the gateway ask the compiler to compile the URI matcher.

Not sure its the cause, but in `GenerateUriMatcher` I can see `g_gateway.clangCompileCodeAndGetFuntions_` is called with one parameter like:

```
clang_engine_addr = 0x0000000000304ad0 {0x0000000000000000}
```

which I guess is not good?

And that parameter can be traced back to

```
UriMatcherCacheEntry* new_entry = GwNewConstructor(UriMatcherCacheEntry);
new_entry->GetClangEngineAddress();
```

which I can't really see how it could be anything but NULL? So maybe this is by design.

**Comment by per-samuelsson** [comment]
Update: @alemoi has been really helpful and we've debugged this on my machine over a debug build of `llvm`. Unfortunally, to no immediate previal. 

It turns out that some method in CLANG return a NULL where we really don't expect any a NULL to be returned. And its not clear why. So I can't really make any more progress here, we'll have to wait and see how this turns out.

**Comment by alemoi** [comment]
Found the reason. 3 months ago I pushed a change to develop Level0 in URI-matcher code: 725e7ae6dc788823cc3742b034f35c198a357fef
Simply didn't push it to pnext, because of many reasons. Anyways, I'm going to push the change to pnext now. And later release more protected scllvm module with more checks.

**Comment by alemoi** [comment]
I wonder why this didn't appear before. Probably because usage of Level1 with pnext is quite recent.

**Comment by alemoi** [comment]
Just pushed the change via 175e157a4155ebbc8dda203d24ad5e73ff2b1487
Let me know if you still have problems.
In general we should start adding automatic tests to pnext branches. Otherwise these kind of issue will appear again and again.

**Comment by per-samuelsson** [comment]
Great @alemoi, now I've totally messed up my environments, but I will sure try this in a while. I'll let you know the result. Thanks.

**Comment by k-rus** [comment]
> In general we should start adding automatic tests to pnext branches.

Good idea. I also think that we need automatic tests on build server. The question is how to make them useful. I created an issue #3043.

**Comment by per-samuelsson** [comment]
> I'll let you know the result. Thanks.

Just realized I forgot to give you the feedback I promise, sorry. When I started fresh, and including your fix, starting the server now works like a charm! :clap: Thanks for all the support, @alemoi.

**Comment by per-samuelsson** [comment]
Made some good progress on this today. I have redesigned bootstrapping of metadata to be driven by code generated code, so

1) we are not limited by static calls when/if we want to expand on number of metadata tables / classes
2) bootstrapping of metadata is faster

```
C:\Users\Per\Git\Starcounter\Develop\Level1\bin\Debug>sccode.exe QUERYPROCESSINGTEST --DatabaseDir=.db --OutputDir=.db.output --TempDir=.db.output --AutoStartExePath="s\QueryProcessingTest\.starcounter\QueryProcessingTest.exe" --FLAG:NoNetworkGateway 
Query processing tests are started
No reload.
Starting direct binding test.
Finished direct binding test.
Test query preparation performance.
First call to prepare query took 88 ms.
Preparing query took 32200 mcs.
Call Prolog (parse and type check) took 2100 mcs.
Map Prolog answer to optimizer input took 10,00 mcs.
Optimizing the query tree took 30,00 mcs.
Finished test query preparation performance.
Test error messages.
Finished test of error messages
Test queries with classes in different namespaces
Finished testing queries with classes in different namespaces
Data population
Finished data population
Start testing queries on conjunction bug (1350)
Finished testing queries on conjunction bug (1350)
Test queries with fetch and order by
Finished test query with fetch and sorting
Test queries with like
Finished test queries with like
Start testing projections
Finished testing projections
Some tests on variables and case insensitivity
Finished some tests on variables and case insensitivity
Test enumerator related bugs
Finished testing enumerator related bugs
Start testing query result mismatch errors.
Finished testing query result mismatch errors.
Test query optimization with indexes
Finished testing query optimization with indexes
Test queries with classes having similar short names
Finished testing queries with classes having similar short names
Test searching object by direct reference lookup.
Finished testing searching object by direct reference lookup.
Testing outer joins
Finished testing outer joins
Test delimited identifiers.
Finidhed testing delimited identifiers.
Test queries with fetch
Finished testing queries with fetch
Test queries with aggregates
Finished testing queries with aggregates
Test code properties
Finished testing code properties
Test expressions in select clause
Finished testing expressions in select clause
Test offset key

Finished testing offset key
Test object identities
Finished testing object identities
Test kinds like behavior
Finished testing kinds like behavior
Benchmark of query cache is skipped
Start unloading query processing database.
All tests completed
```

Some refinements left and checking that I've covered all code, but I should be able to push this either later tonight or at least tomorrow.

**Comment by k-rus** [comment]
Great!
Passing the test to this level is enough. Don't forget to merge latest pnext, I enabled more tests in QueryProcessingTest.

I am wonder should call to metalayer to create meta-tables and populate meta-data (function `PopulateRuntimeMetadata`) be earlier during bootstrap? 
The only requirement is that context (kernel function `star_get_context`) is created before calling it.

**Comment by per-samuelsson** [comment]
> Don't forget to merge latest pnext, I enabled more tests in QueryProcessingTest.

I did that today during the day, but I'll of course merge again before the final test, I always do.

> I am wonder should call to metalayer to create meta-tables and populate meta-data (function PopulateRuntimeMetadata) be earlier during bootstrap?

Wow. That I don't know, sorry. What has changed to drive such change? I really dont have the big picture to give an accurate answer. Maybe we can try discuss that and see?

**Comment by k-rus** [comment]
> I am wonder should call to metalayer to create meta-tables and populate meta-data (function PopulateRuntimeMetadata) be earlier during bootstrap?
> Wow. That I don't know, sorry. What has changed to drive such change? I really dont have the big picture to give an accurate answer. Maybe we can try discuss that and see?

Current place (if you haven't changed) is unsafe, since it depends if the call for base package or user-defined classes. The condition checks if table name is the first meta-table in the json list and if the first meta-table is changed, the condition has to be changed too.

**Comment by per-samuelsson** [comment]
> The condition checks if table name is the first meta-table in the json list and if the first meta-table is changed, the condition has to be changed too.

Well, I've already redesigned that, b/c that kind of check seemed quite ... well, not really safe. I've made it a little more OOP, so that the starcounter package logic is in its own package class (`StarcounterPackage`).

For initializing of type specifications, for example, the call in `UpdateDatabaseSchemaAndRegisterTypes` to `InitTypeSpecifications` is now virtual. The standard package define no special logic, while `StarcounterPackage` does this:

```C#
protected override void InitTypeSpecifications() {
            // User-level classes are self registering and report in to
            // the installed host manager on first use (via an emitted call
            // in the static class constructor). For system classes, we
            // have to do this by hand.

            var metaBinder = MetaBinder.Instance;
            foreach (var type in metaBinder.GetSpecifications()) {
                HostManager.InitTypeSpecification(type);
            }
        }
```

**Comment by per-samuelsson** [comment]
BTW, the idea now is I merge this `pnextmerge2728` branch into `pnext`? Or is it better you do that? Its up to date with latest `pnext`, so its ready in that sense.

**Comment by k-rus** [comment]
> BTW, the idea now is I merge this pnextmerge2728 branch into pnext? Or is it better you do that? Its up to date with latest pnext, so its ready in that sense.

You can merge it, if you haven't moved from it. We need it in pnext.

**Comment by per-samuelsson** [comment]
> You can merge it, if you haven't moved from it. We need it in pnext.

OK, I will when you clear that one last time. Feels a bit shaky, since I can't do the normal personal build + testing, etc. But if you say its fine, I'll do it. No probs.

Just to be clear, with the latest `pnext` merged into this branch, the point I'm at now is that the first run of the query processing test succeed:

```
Query processing tests are started
No reload.
Starting direct binding test.
Finished direct binding test.
Test query preparation performance.
First call to prepare query took 79 ms.
Preparing query took 1900 mcs.
Call Prolog (parse and type check) took 2000 mcs.
Map Prolog answer to optimizer input took 10,00 mcs.
Optimizing the query tree took 30,00 mcs.
Finished test query preparation performance.
Test error messages.
Finished test of error messages
Test queries with classes in different namespaces
Finished testing queries with classes in different namespaces
Data population
Finished data population
Start testing queries on conjunction bug (1350)
Finished testing queries on conjunction bug (1350)
Test queries with fetch and order by
Finished test query with fetch and sorting
Test queries with like
Finished test queries with like
Start testing projections
Finished testing projections
Some tests on variables and case insensitivity
Finished some tests on variables and case insensitivity
Test enumerator related bugs
Finished testing enumerator related bugs
Start testing query result mismatch errors.
Finished testing query result mismatch errors.
Test query optimization with indexes
Finished testing query optimization with indexes
Test queries with classes having similar short names
Finished testing queries with classes having similar short names
Test searching object by direct reference lookup.
Finished testing searching object by direct reference lookup.
Testing outer joins
Finished testing outer joins
Test delimited identifiers.
Finidhed testing delimited identifiers.
Test queries with fetch
Finished testing queries with fetch
Test queries with aggregates
Finished testing queries with aggregates
Test code properties
Finished testing code properties
Test expressions in select clause
Finished testing expressions in select clause
Test offset key

Finished testing offset key
Test object identities
Finished testing object identities
Test kinds like behavior
Finished testing kinds like behavior
Benchmark of query cache is skipped
Start unloading query processing database.
All tests completed
```

And the second run fails on this:
```
System.EntryPointNotFoundException: Unable to find an entry point named 'star_create_index' in DLL 'sccoredb.dll'.
   at Starcounter.Internal.sccoredb.star_create_index(UInt32 transaction_flags, UInt16 table_id, String name, UInt16 sort_mask, Int16* column_indexes, UInt32 flags)
   at Starcounter.Binding.TableUpgrade.MoveIndexesToNewTable() in C:\Users\Per\Git\Starcounter\Develop\Level1\src\Starcounter\Binding\TableUpgrade.cs:line 419
   at Starcounter.Binding.TableUpgrade.Eval() in C:\Users\Per\Git\Starcounter\Develop\Level1\src\Starcounter\Binding\TableUpgrade.cs:line 92
   at Starcounter.Hosting.Package.CreateOrUpdateDatabaseTable(TypeDef typeDef) in C:\Users\Per\Git\Starcounter\Develop\Level1\src\Hosting\Starcounter.Hosting\Package.cs:line 464
   at Starcounter.Hosting.Package.<>c__DisplayClass20_1.<UpdateDatabaseSchemaAndRegisterTypes>b__2() in C:\Users\Per\Git\Starcounter\Develop\Level1\src\Hosting\Starcounter.Hosting\Package.cs:line 348
   at Starcounter.Internal.TransactionManager.Scope(TransactionHandle handle, Action action) in C:\Users\Per\Git\Starcounter\Develop\Level1\src\Starcounter\TransactionManager.cs:line 464
   at Starcounter.Transaction.Scope(Action action) in C:\Users\Per\Git\Starcounter\Develop\Level1\src\Starcounter\Transaction.cs:line 199
   at Starcounter.Hosting.Package.UpdateDatabaseSchemaAndRegisterTypes(String fullAppId, TypeDef[] unregisteredTypeDefs, TypeDef[] allTypeDefs) in C:\Users\Per\Git\Starcounter\Develop\Level1\src\Hosting\Starcounter.Hosting\Package.cs:line 343
   at Starcounter.Hosting.Package.ProcessWithinCurrentApplication(Application application) in C:\Users\Per\Git\Starcounter\Develop\Level1\src\Hosting\Starcounter.Hosting\Package.cs:line 250
   at Starcounter.Hosting.Package.Process() in C:\Users\Per\Git\Starcounter\Develop\Level1\src\Hosting\Starcounter.Hosting\Package.cs:line 143
   at Starcounter.Hosting.Package.Process(IntPtr hPackage) in C:\Users\Per\Git\Starcounter\Develop\Level1\src\Hosting\Starcounter.Hosting\Package.cs:line 55
   at StarcounterInternal.Hosting.Processor.RunMessageLoop(Void* hsched) in C:\Users\Per\Git\Starcounter\Develop\Level1\src\Hosting\Starcounter.Hosting\Processor.cs:line 78
TypeName=
HResult=-2146233053
```

If you say I can merge like that, I'll do so right away and then switch back to something else.

**Comment by per-samuelsson** [comment]
> If you say I can merge like that, I'll do so right away and then switch back to something else.

@k-rus cleared this merge, so now I've merged this whole effort into **pnext** in https://github.com/Starcounter/Starcounter/commit/d91220b96632239efda8c8b081cb7774bb1eb6e5. This merge also include the former pending merge of #2728.

Let me know how this turns out @k-rus, and close the issue once verified. Thanks.

**Comment by k-rus** [comment]
I am trying to build Level1 with this change and I fail with:
```
5>  C:\GitRepositories\Starcounter\Level1\bin\Debug\metaschema.json -> C:\GitRepositories\Starcounter\Level1\src\Starcounter\Metadata
5>C:\GitRepositories\Starcounter\Level1\src\Starcounter\Metadata\GeneratedMetadataClasses.cs(332,28,332,59): error CS0109: The member 'Token.__starcounterTypeSpecification' does not hide an inherited member. The new keyword is not required.
5>C:\GitRepositories\Starcounter\Level1\src\Starcounter\Metadata\GeneratedMetadataClasses.cs(351,28,351,59): error CS0109: The member 'SetSpecifier.__starcounterTypeSpecification' does not hide an inherited member. The new keyword is not required.
5>C:\GitRepositories\Starcounter\Level1\src\Starcounter\Metadata\GeneratedMetadataClasses.cs(30,28,30,59): error CS0109: The member 'Type.__starcounterTypeSpecification' does not hide an inherited member. The new keyword is not required.
5>C:\GitRepositories\Starcounter\Level1\src\Starcounter\Metadata\GeneratedMetadataClasses.cs(185,28,185,59): error CS0109: The member 'Member.__starcounterTypeSpecification' does not hide an inherited member. The new keyword is not required.
5>C:\GitRepositories\Starcounter\Level1\src\Starcounter\Metadata\GeneratedMetadataClasses.cs(280,28,280,59): error CS0109: The member 'Index.__starcounterTypeSpecification' does not hide an inherited member. The new keyword is not required.
5>C:\GitRepositories\Starcounter\Level1\src\Starcounter\Metadata\GeneratedMetadataClasses.cs(303,28,303,59): error CS0109: The member 'IndexedColumn.__starcounterTypeSpecification' does not hide an inherited member. The new keyword is not required.
```

**Comment by per-samuelsson** [comment]
> I am trying to build Level1 with this change and I fail

Hmm, thats bad. So maybe I shouldn't have merged after all. Looks there is some diff in "treat warnings as errors", maybe my settings were softer for some reason (although I haven't changed that on purpose for sure).

Can you try just relax that as a test? Under Project Properties. Then I'll remove the warnings meanwhile.

**Comment by k-rus** [comment]
Does not help. It is really errors (not warnings treated as errors). An interesting thing that I didn't have problem to build #2728 some time ago.

**Comment by k-rus** [comment]
> Does not help. It is really errors (not warnings treated as errors). An interesting thing that I didn't have problem to build #2728 some time ago.

Rebuild of exactly this project with not treating warnings as errors worked.

**Comment by per-samuelsson** [comment]
> Does not help. It is really errors (not warnings treated as errors)

[Compiler Warning (level 4) CS0109](https://msdn.microsoft.com/en-us/library/css4y2c4(v=vs.90).aspx)

> An interesting thing that I didn't have problem to build #2728 some time ago.

Yeah, but I decided to change generation to support binding via generated code, not reflection, for a nicer bootstrapping that was also faster.

**Comment by k-rus** [comment]
Finally, I found where I should do it. I set Treat warnings as errors to none on Starcounter project and build succeeded.

**Comment by k-rus** [comment]
After this, QueryProcessingTest works as expected! So it is only left to fix warnings.

**Comment by per-samuelsson** [comment]
> So it is only left to fix warnings.

Pushed. Sorry for the inconvenience.

Turn the strictness on again, test and close this when confirmed it works. Thanks.

**Comment by k-rus** [comment]
Confirm the fix. Closing.
Thank you!
